File: Solutions/C++/Q3.cpp
#include<bits/stdc++.h>

using namespace std;


int indexOf(int * arr, int elem, int i, int len) {
	while (i < len) {
		if (arr[i] == elem)
			return i;
		i++;
	}

  return -1;
}

string findPair(int * arr, int n, int k) {

	double lowestAbs = 1000000;
	string lowestFound = "-1";

	int j;
	int min;
	
	for (int i = 0; i < n-1; i++) {
		j = indexOf(arr, arr[i] + k, i, n);

		if (j != -1 && i <= j) {
			min = abs(i+1 - j+1);

			if (lowestAbs > min) {
				lowestFound = to_string(i+1) + " " + to_string(j+1);
				lowestAbs = min;
			}
		}
	}

	return lowestFound;
}

string find_min_days(int numOfDays, int numOfProfits, int dailyShareValue[], int profitsRequired[]) {
	string str = "";
	for (int i=0; i<numOfProfits; i++) {
		if (i != 0)
			str += ",";
		str += findPair(dailyShareValue, numOfDays, profitsRequired[i]);
	}
  return str;
}

int main() {
	int numOfDays,numOfProfits,i;
	string answer="";
	cin>>numOfDays>>numOfProfits;
	int dailyShareValue[numOfDays];
	int profitsRequired[numOfProfits];
	for (i=0;i<numOfDays;i++)
		cin>>dailyShareValue[i];
	for (i=0;i<numOfProfits;i++)
    cin>>profitsRequired[i];

  answer = find_min_days(numOfDays, numOfProfits, dailyShareValue, profitsRequired);

  // Do not remove below line
	cout<<answer<<endl;
  // Do not print anything after this line

	return 0;
}

File: Solutions/TypeScript (Base)/Q3.ts
type Input = string[];

const Q3Func = (input: Input): string => {
	const [numOfDays, numOfProfits] = input[0].split(' ').map(v => parseInt(v));
	const dailyShareValue: number[] = input[1].split(' ').map(v => parseInt(v));
	const profitsRequired: number[] = input.slice(2).map(v => parseInt(v));

  return profitsRequired.map(v => {
		return findPair(dailyShareValue, numOfDays, v);
	}).join(',');
};

export default Q3Func;

function indexOf(arr: number[], elem: number, i: number): number {
	let len = arr.length;

	while (i < len) {
		if (arr[i] == elem)
			return i;
		i++;
	}

  return -1;
}

export function findPair(arr: number[], n: number, k: number): string {
	let lowestAbs: number = 1000000;
	let lowestFound: string = "-1";

	let j: number;
	let min: number;

	for (let i = 0; i < n-1; i++) {
		j = indexOf(arr, arr[i] + k, i);

		if (j != -1 && i <= j) {
			min = Math.abs(i+1 - j+1);

			if (lowestAbs > min) {
				lowestFound = `${i+1} ${j+1}`;
				lowestAbs = min;
			}
		}
	}

	return lowestFound;
}